tests/all/gc.rs

```diff
@@ -0,0 +1,228 @@
+    config.wasm_reference_types(true);
+
+    let engine = Engine::new(&config);
+    let store = Store::new(&engine);
+
+    let module = Module::new(&engine, source)?;
+
+    Ok((store, module))
+}
+
+#[test]
+fn smoke_test_gc() -> anyhow::Result<()> {
+    let (store, module) = ref_types_module(
+        r#"
+            (module
+                (import "" "" (func $do_gc))
+                (func $recursive (export "func") (param i32 externref) (result externref)
+                    local.get 0
+                    i32.eqz
+                    if (result externref)
+                        call $do_gc
+                        local.get 1
+                    else
+                        local.get 0
+                        i32.const 1
+                        i32.sub
+                        local.get 1
+                        call $recursive
+                    end
+                )
+            )
+        "#,
+    )?;
+
+    let do_gc = Func::wrap(&store, {
+        let store = store.clone();
+        move || {
+            // Do a GC with `externref`s on the stack in Wasm frames.
+            store.gc();
+        }
+    });
+    let instance = Instance::new(&store, &module, &[do_gc.into()])?;
+    let func = instance.get_func("func").unwrap();
+
+    let inner_dropped = Rc::new(Cell::new(false));
+    let r = ExternRef::new(&store, SetFlagOnDrop(inner_dropped.clone()));
+    {
+        let args = [Val::I32(5), Val::ExternRef(Some(r.clone()))];
+        func.call(&args)?;
+    }
+
+    // Still held alive by the `VMExternRefActivationsTable` (potentially in
+    // multiple slots within the table) and by this `r` local.
+    assert!(r.get_reference_count() >= 2);
+
+    // Doing a GC should see that there aren't any `externref`s on the stack in
+    // Wasm frames anymore.
+    store.gc();
+    assert_eq!(r.get_reference_count(), 1);
+
+    // Dropping `r` should drop the inner `SetFlagOnDrop` value.
+    drop(r);
+    assert!(inner_dropped.get());
+
+    return Ok(());
+
+    struct SetFlagOnDrop(Rc<Cell<bool>>);
+
+    impl Drop for SetFlagOnDrop {
+        fn drop(&mut self) {
+            self.0.set(true);
+        }
+    }
+}
+
+#[test]
+fn wasm_dropping_refs() -> anyhow::Result<()> {
+    let (store, module) = ref_types_module(
+        r#"
+            (module
+                (func (export "drop_ref") (param externref)
+                    nop
+                )
+            )
+        "#,
+    )?;
+
+    let instance = Instance::new(&store, &module, &[])?;
+    let drop_ref = instance.get_func("drop_ref").unwrap();
+
+    let num_refs_dropped = Rc::new(Cell::new(0));
+
+    // NB: 4096 is greater than the initial `VMExternRefActivationsTable`
+    // capacity, so this will trigger at least one GC.
+    for _ in 0..4096 {
+        let r = ExternRef::new(&store, CountDrops(num_refs_dropped.clone()));
+        let args = [Val::ExternRef(Some(r))];
+        drop_ref.call(&args)?;
+    }
+
+    assert!(num_refs_dropped.get() > 0);
```

Comment:
Answering my own question, if my thought earlier about the stack canary and rvalue promotion breaks things this assertion will break. If our stack canary *never* works then nothing will get gc'd from the above loop.

Reply:
Ah, yep, I see you figured it out :)